<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  Top Level Namespace
  
    &mdash; Documentation by YARD 0.9.25
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index</a> &raquo;
    
    
    <span class="title">Top Level Namespace</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><h1>Top Level Namespace
  
  
  
</h1>
<div class="box_info">
  

  
  
  
  
  

  

  
</div>








  
    <h2>
      Instance Method Summary
      <small><a href="#" class="summary_toggle">collapse</a></small>
    </h2>

    <ul class="summary">
      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#create_and_add_features-instance_method" title="#create_and_add_features (instance method)">#<strong>create_and_add_features</strong>(biosequence, gene_id, positions_array, strand)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>For every pair of positions in a position array, this function adds a new
&#39;CTTCTT_direct_repeat&#39; feature to a Bio::Sequence object,
containing the position, sequence, strand, and an unique ID.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#fetch-instance_method" title="#fetch (instance method)">#<strong>fetch</strong>(url, headers = {accept: &quot;*/*&quot;}, user = &quot;&quot;, pass = &quot;&quot;)  &#x21d2; String, false </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Access an URL, function donated by Mark Wilkinson.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#find_seq_in_exons-instance_method" title="#find_seq_in_exons (instance method)">#<strong>find_seq_in_exons</strong>(embl_hash)  &#x21d2; Hash&lt;Symbol, Bio::Sequence&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Given a hash of id =&gt; Bio::EMBL object, it finds every
&#39;CTTCTT&#39; sequence in the exons of each gene, adds that feature to a
new Bio::Sequence object containing all the previous information of the
Bio::EMBL object, and returns the new hash of id =&gt; Bio::Sequence
object.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#get_embl-instance_method" title="#get_embl (instance method)">#<strong>get_embl</strong>(gene_array)  &#x21d2; Hash&lt;Symbol, Bio::EMBL&gt; </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Given an array of gene ids, it gets the EMBL entry from ensemblgenomesgene,
creates a Bio::EMBL object and stores it into a hash.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#read_gene_list-instance_method" title="#read_gene_list (instance method)">#<strong>read_gene_list</strong>(path)  &#x21d2; Array&lt;Symbol&gt;<sup>?</sup> </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Reads a file containing one gene id (in AGI locus code format) in each
column and stores each id as a lowercase symbol on an array and returns it.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_gff3-instance_method" title="#write_gff3 (instance method)">#<strong>write_gff3</strong>(new_embl_hash, mode, filename)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Writes a GFF3 file with all of the &#39;CTTCTT_direct_repeat&#39; features.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_gff3_global-instance_method" title="#write_gff3_global (instance method)">#<strong>write_gff3_global</strong>(new_embl_hash, filename = &quot;CTTCTT_GFF3_chromosome.gff&quot;)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Writes a GFF3 file with all of the &#39;CTTCTT_direct_repeat&#39; features,
using the start of the chromosome as the coordinate origin.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_gff3_local-instance_method" title="#write_gff3_local (instance method)">#<strong>write_gff3_local</strong>(new_embl_hash, filename = &quot;CTTCTT_GFF3_gene.gff&quot;)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Writes a GFF3 file with all of the &#39;CTTCTT_direct_repeat&#39; features,
using the start of the gene as the coordinate origin.</p>
</div></span>
  
</li>

      
        <li class="public ">
  <span class="summary_signature">
    
      <a href="top-level-namespace.html#write_report-instance_method" title="#write_report (instance method)">#<strong>write_report</strong>(gene_array, new_embl_hash, filename = &quot;CTTCTT_report.txt&quot;)  &#x21d2; Object </a>
    

    
  </span>
  
  
  
  
  
  
  

  
    <span class="summary_desc"><div class='inline'>
<p>Writes a report with the IDs of all the genes in the initial gene list that
don&#39;t contain &#39;CTTCTT&#39; in any of their exons.</p>
</div></span>
  
</li>

      
    </ul>
  



  <div id="instance_method_details" class="method_details_list">
    <h2>Instance Method Details</h2>

    
      <div class="method_details first">
  <h3 class="signature first" id="create_and_add_features-instance_method">
  
    #<strong>create_and_add_features</strong>(biosequence, gene_id, positions_array, strand)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>For every pair of positions in a position array, this function adds a new
&#39;CTTCTT_direct_repeat&#39; feature to a Bio::Sequence object,
containing the position, sequence, strand, and an unique ID.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>biosequence</span>
      
      
        <span class='type'>(<tt>Bio::Sequence</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The Bio::Sequence object, made from a Bio::EMBL object to be able to add
features to it</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>gene_id</span>
      
      
        <span class='type'>(<tt>Symbol</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The id of the gene</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>positions_array</span>
      
      
        <span class='type'>(<tt>Array&lt;Array&lt;Integer&gt;&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>An array of arrays, each subarray representing start and end positions,
1-indexed, of the feature in the gene</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>strand</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>“+” for forward strand, “-” for the reverse strand</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 105</span>

<span class='kw'>def</span> <span class='id identifier rubyid_create_and_add_features'>create_and_add_features</span><span class='lparen'>(</span><span class='id identifier rubyid_biosequence'>biosequence</span><span class='comma'>,</span> <span class='id identifier rubyid_gene_id'>gene_id</span><span class='comma'>,</span> <span class='id identifier rubyid_positions_array'>positions_array</span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='rparen'>)</span>
    
    <span class='id identifier rubyid_positions_array'>positions_array</span><span class='period'>.</span><span class='id identifier rubyid_uniq'>uniq</span><span class='period'>.</span><span class='id identifier rubyid_each_with_index'>each_with_index</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_position_pair'>position_pair</span><span class='comma'>,</span> <span class='id identifier rubyid_index'>index</span><span class='op'>|</span>
        <span class='id identifier rubyid_start_pos'>start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='op'>*</span><span class='id identifier rubyid_position_pair'>position_pair</span>
        <span class='comment'># creating the position string
</span>        <span class='kw'>if</span> <span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='int'>+1</span> <span class='op'>||</span> <span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+</span><span class='tstring_end'>&quot;</span></span> <span class='op'>||</span> <span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+1</span><span class='tstring_end'>&quot;</span></span>
            <span class='id identifier rubyid_position_string'>position_string</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_pos'>start_pos</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_pos'>end_pos</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>
            <span class='id identifier rubyid_strand_label'>strand_label</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+</span><span class='tstring_end'>&quot;</span></span>
        <span class='kw'>elsif</span> <span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='op'>-</span><span class='int'>1</span> <span class='op'>||</span> <span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-</span><span class='tstring_end'>&quot;</span></span> <span class='op'>||</span> <span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-1</span><span class='tstring_end'>&quot;</span></span>
            <span class='id identifier rubyid_position_string'>position_string</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>complement(</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_start_pos'>start_pos</span><span class='embexpr_end'>}</span><span class='tstring_content'>..</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_end_pos'>end_pos</span><span class='embexpr_end'>}</span><span class='tstring_content'>)</span><span class='tstring_end'>&quot;</span></span>
            <span class='id identifier rubyid_strand_label'>strand_label</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-</span><span class='tstring_end'>&quot;</span></span>
        <span class='kw'>else</span>
            <span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>WARNING: For </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='embexpr_end'>}</span><span class='tstring_content'>, the strand parameter was </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand'>strand</span><span class='embexpr_end'>}</span><span class='tstring_content'>. It should be &#39;+1&#39; or &#39;+&#39; for the forward, and &#39;-1&#39; or &#39;-&#39; for the reverse</span><span class='tstring_end'>&quot;</span></span>
            <span class='kw'>next</span>
        <span class='kw'>end</span>

        <span class='id identifier rubyid_new_feature'>new_feature</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_direct_repeat</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_position'>position</span> <span class='op'>=</span> <span class='id identifier rubyid_position_string'>position_string</span><span class='comma'>,</span> 
                                        <span class='id identifier rubyid_qualifiers'>qualifiers</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>sequence</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>CTTCTT</span><span class='tstring_end'>&#39;</span></span><span class='rparen'>)</span><span class='comma'>,</span>
                                                        <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_strand_label'>strand_label</span><span class='rparen'>)</span><span class='comma'>,</span>
                                                        <span class='const'>Bio</span><span class='op'>::</span><span class='const'>Feature</span><span class='op'>::</span><span class='const'>Qualifier</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&#39;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&#39;</span></span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span><span class='embexpr_end'>}</span><span class='tstring_content'>.CTTCTT_repeat.</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_strand_label'>strand_label</span><span class='embexpr_end'>}</span><span class='tstring_content'>.</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_index'>index</span> <span class='op'>+</span> <span class='int'>1</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rbracket'>]</span><span class='rparen'>)</span>
                                                        <span class='comment'># id, for example: AT2G46340.CTTCTT_repeat.-.1
</span>                                                        <span class='comment'># that way I can use it afterwards for the GFF3 file
</span>        <span class='id identifier rubyid_biosequence'>biosequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span> <span class='op'>&lt;&lt;</span> <span class='id identifier rubyid_new_feature'>new_feature</span>
    <span class='kw'>end</span>
    <span class='comment'># Don&#39;t need to return anything because the new feature is being added to the Bio::Sequence object
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="fetch-instance_method">
  
    #<strong>fetch</strong>(url, headers = {accept: &quot;*/*&quot;}, user = &quot;&quot;, pass = &quot;&quot;)  &#x21d2; <tt>String</tt>, <tt>false</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Access an URL, function donated by Mark Wilkinson.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>url</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>URL</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>headers</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>{accept: &quot;*/*&quot;}</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>headers</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>user</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>username, default “”</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>pass</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>password, default “”</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>String</tt>, <tt>false</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>the contents of the URL, or false if an exception occured</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 14</span>

<span class='kw'>def</span> <span class='id identifier rubyid_fetch'>fetch</span><span class='lparen'>(</span><span class='id identifier rubyid_url'>url</span><span class='comma'>,</span> <span class='id identifier rubyid_headers'>headers</span> <span class='op'>=</span> <span class='lbrace'>{</span><span class='label'>accept:</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>*/*</span><span class='tstring_end'>&quot;</span></span><span class='rbrace'>}</span><span class='comma'>,</span> <span class='id identifier rubyid_user'>user</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_pass'>pass</span><span class='op'>=</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Request</span><span class='period'>.</span><span class='id identifier rubyid_execute'>execute</span><span class='lparen'>(</span><span class='lbrace'>{</span>
        <span class='label'>method:</span> <span class='symbol'>:get</span><span class='comma'>,</span>
        <span class='label'>url:</span> <span class='id identifier rubyid_url'>url</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='comma'>,</span>
        <span class='label'>user:</span> <span class='id identifier rubyid_user'>user</span><span class='comma'>,</span>
        <span class='label'>password:</span> <span class='id identifier rubyid_pass'>pass</span><span class='comma'>,</span>
        <span class='label'>headers:</span> <span class='id identifier rubyid_headers'>headers</span><span class='rbrace'>}</span><span class='rparen'>)</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span>
    
    <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>ExceptionWithResponse</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span> 
    <span class='kw'>rescue</span> <span class='const'>RestClient</span><span class='op'>::</span><span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span>
        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span> 
    <span class='kw'>rescue</span> <span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span> <span class='comment'># generic
</span>        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='kw'>false</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_response'>response</span> 
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="find_seq_in_exons-instance_method">
  
    #<strong>find_seq_in_exons</strong>(embl_hash)  &#x21d2; <tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Given a hash of id =&gt; Bio::EMBL object, it finds every
&#39;CTTCTT&#39; sequence in the exons of each gene, adds that feature to a
new Bio::Sequence object containing all the previous information of the
Bio::EMBL object, and returns the new hash of id =&gt; Bio::Sequence
object.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>embl_hash</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::EMBL&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The hash of Bio::EMBL objects for each EMBL entry</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The new hash of Bio::Sequence objects, containing only the ids of those
genes with &#39;CTTCTT&#39; in any of their exons</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 157</span>

<span class='kw'>def</span> <span class='id identifier rubyid_find_seq_in_exons'>find_seq_in_exons</span><span class='lparen'>(</span><span class='id identifier rubyid_embl_hash'>embl_hash</span><span class='rparen'>)</span>

    <span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span> <span class='op'>=</span> <span class='lbrace'>{</span><span class='rbrace'>}</span> <span class='comment'># Hash to store {gene_id =&gt; Bio::EMBL object with the new features}
</span>
    <span class='id identifier rubyid_embl_hash'>embl_hash</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='comma'>,</span> <span class='id identifier rubyid_bio_embl'>bio_embl</span><span class='op'>|</span> <span class='comment'># iterating over the hash
</span>
        <span class='id identifier rubyid_bio_embl_as_biosequence'>bio_embl_as_biosequence</span> <span class='op'>=</span> <span class='id identifier rubyid_bio_embl'>bio_embl</span><span class='period'>.</span><span class='id identifier rubyid_to_biosequence'>to_biosequence</span> <span class='comment'># turning the Bio::EMBL into Bio::Sequence so that I can add features
</span>        <span class='id identifier rubyid_embl_seq'>embl_seq</span> <span class='op'>=</span> <span class='id identifier rubyid_bio_embl'>bio_embl</span><span class='period'>.</span><span class='id identifier rubyid_seq'>seq</span> <span class='comment'># getting the dna sequence as Bio::Sequence::NA
</span>
        <span class='id identifier rubyid_all_positions_forward'>all_positions_forward</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span> <span class='comment'># store the positions of the feature CTTCTT on the forward strand
</span>        <span class='id identifier rubyid_all_positions_reverse'>all_positions_reverse</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span> <span class='comment'># store the positions of the feature CTTCTT on the reverse strand
</span>
        <span class='comment'># iterating through each feature of the original bio_embl
</span>        <span class='comment'># not using bio_embl_as_biosequence here because I will be adding features 
</span>        <span class='comment'># and it&#39;s not good to manipulate an object while iterating through it
</span>        <span class='id identifier rubyid_bio_embl'>bio_embl</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span> 
            <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>exon</span><span class='tstring_end'>&quot;</span></span> <span class='comment'># skip if the feature is not an exon
</span>            <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>note</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_match'>match</span><span class='lparen'>(</span><span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>i</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='rparen'>)</span> <span class='comment'># skip if the exon is from another gene (because of overlapping or whatever)
</span>
            <span class='id identifier rubyid_bio_location'>bio_location</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span> <span class='comment'># feature.locations is Bio::Locations, which is basically an array of Bio::Location objects
</span>            <span class='id identifier rubyid_exon_start_pos'>exon_start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_exon_end_pos'>exon_end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='comma'>,</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span> <span class='comment'># 1-indexed, position of the exon with respect to the start of the gene
</span>
            <span class='id identifier rubyid_exon_seq'>exon_seq</span> <span class='op'>=</span> <span class='id identifier rubyid_embl_seq'>embl_seq</span><span class='period'>.</span><span class='id identifier rubyid_subseq'>subseq</span><span class='lparen'>(</span><span class='id identifier rubyid_exon_start_pos'>exon_start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_exon_end_pos'>exon_end_pos</span><span class='rparen'>)</span> <span class='comment'># getting the sequence of the exon, 1-indexed, forward strand
</span>
            <span class='kw'>if</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='int'>+1</span> <span class='comment'># the feature is on the current strand: we need to search for CTTCTT in the forward strand
</span>                <span class='comment'># To find the sequence, I&#39;m using lookahead (?=) so that I can match overlapping expressions 
</span>                <span class='comment'># (https://stackoverflow.com/questions/41028602/find-overlapping-regexp-matches)
</span>                <span class='comment'># (https://stackoverflow.com/questions/5241653/ruby-regex-match-and-get-positions-of)
</span>                <span class='id identifier rubyid_start_f'>start_f</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=(cttctt))</span><span class='regexp_end'>/i</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span> <span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_begin'>begin</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='int'>1</span><span class='rbrace'>}</span> <span class='comment'># +1 so it is 1-indexed
</span>                
                <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_start_f'>start_f</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># if there is no match, skip
</span>                <span class='comment'># if there is a match
</span>                <span class='comment'># Positions: 
</span>                <span class='comment'># pos is the position of the feature with respect to the exon
</span>                <span class='comment'># to get it with respect to the gene, we have to sum pos + exon_start_pos
</span>                <span class='comment'># everything is 1-indexed, so when summing two 1-indexed positions we have to substract 1.
</span>                <span class='id identifier rubyid_positions_f'>positions_f</span> <span class='op'>=</span> <span class='id identifier rubyid_start_f'>start_f</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span><span class='op'>|</span><span class='id identifier rubyid_pos'>pos</span><span class='op'>|</span> <span class='lbracket'>[</span><span class='id identifier rubyid_pos'>pos</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start_pos'>exon_start_pos</span> <span class='op'>-</span> <span class='int'>1</span><span class='comma'>,</span> <span class='id identifier rubyid_pos'>pos</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start_pos'>exon_start_pos</span> <span class='op'>-</span><span class='int'>1</span> <span class='op'>+</span> <span class='int'>5</span><span class='rbracket'>]</span><span class='rbrace'>}</span> <span class='kw'>unless</span> <span class='id identifier rubyid_start_f'>start_f</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># 1-indexed
</span>                <span class='id identifier rubyid_all_positions_forward'>all_positions_forward</span> <span class='op'>|=</span> <span class='id identifier rubyid_positions_f'>positions_f</span> <span class='comment'># include the [start_pos, end_pos] pairs that aren&#39;t included already
</span>
            <span class='kw'>elsif</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_strand'>strand</span> <span class='op'>==</span> <span class='op'>-</span><span class='int'>1</span> <span class='comment'># the feature is on the complement strand: we need to search for the reverse complement, AAGAAG, in the forward strand
</span>                <span class='comment'># To find the sequence, I&#39;m using lookahead (?=) so that I can match overlapping expressions 
</span>                <span class='comment'># (https://stackoverflow.com/questions/41028602/find-overlapping-regexp-matches)
</span>                <span class='comment'># (https://stackoverflow.com/questions/5241653/ruby-regex-match-and-get-positions-of)
</span>                <span class='id identifier rubyid_start_r'>start_r</span> <span class='op'>=</span> <span class='id identifier rubyid_exon_seq'>exon_seq</span><span class='period'>.</span><span class='id identifier rubyid_enum_for'>enum_for</span><span class='lparen'>(</span><span class='symbol'>:scan</span><span class='comma'>,</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>(?=(aagaag))</span><span class='regexp_end'>/i</span></span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span> <span class='const'>Regexp</span><span class='period'>.</span><span class='id identifier rubyid_last_match'>last_match</span><span class='period'>.</span><span class='id identifier rubyid_begin'>begin</span><span class='lparen'>(</span><span class='int'>0</span><span class='rparen'>)</span> <span class='op'>+</span> <span class='int'>1</span><span class='rbrace'>}</span> <span class='comment'># 1-indexed
</span>
                <span class='kw'>next</span> <span class='kw'>if</span> <span class='id identifier rubyid_start_r'>start_r</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># if there is no match, skip
</span>                <span class='comment'># if there is a match
</span>                <span class='comment'># Positions: 
</span>                <span class='comment'># pos is the position of the feature with respect to the exon
</span>                <span class='comment'># to get it with respect to the gene, we have to sum pos + exon_start_pos
</span>                <span class='comment'># everything is 1-indexed, so when summing two 1-indexed positions we have to substract 1.
</span>                <span class='id identifier rubyid_positions_r'>positions_r</span> <span class='op'>=</span> <span class='id identifier rubyid_start_r'>start_r</span><span class='period'>.</span><span class='id identifier rubyid_map'>map</span> <span class='lbrace'>{</span><span class='op'>|</span><span class='id identifier rubyid_pos'>pos</span><span class='op'>|</span> <span class='lbracket'>[</span><span class='id identifier rubyid_pos'>pos</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start_pos'>exon_start_pos</span> <span class='op'>-</span> <span class='int'>1</span><span class='comma'>,</span> <span class='id identifier rubyid_pos'>pos</span> <span class='op'>+</span> <span class='id identifier rubyid_exon_start_pos'>exon_start_pos</span> <span class='op'>-</span><span class='int'>1</span> <span class='op'>+</span> <span class='int'>5</span><span class='rbracket'>]</span><span class='rbrace'>}</span> <span class='kw'>unless</span> <span class='id identifier rubyid_start_r'>start_r</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># 1-indexed
</span>                <span class='id identifier rubyid_all_positions_reverse'>all_positions_reverse</span> <span class='op'>|=</span> <span class='id identifier rubyid_positions_r'>positions_r</span> <span class='comment'># include the [start_pos, end_pos] pairs that aren&#39;t included already
</span>            <span class='kw'>else</span>
                <span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>WARNING: no strand</span><span class='tstring_end'>&quot;</span></span>
                <span class='kw'>next</span> <span class='comment'># skip
</span>            <span class='kw'>end</span>
        <span class='kw'>end</span>
        <span class='comment'># once every feature has been scanned for that Bio::EMBL object, we add the new CTTCTT features
</span>        <span class='id identifier rubyid_create_and_add_features'><span class='object_link'><a href="#create_and_add_features-instance_method" title="#create_and_add_features (method)">create_and_add_features</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_bio_embl_as_biosequence'>bio_embl_as_biosequence</span><span class='comma'>,</span> <span class='id identifier rubyid_gene_id'>gene_id</span><span class='comma'>,</span> <span class='id identifier rubyid_all_positions_forward'>all_positions_forward</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>+</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
        <span class='id identifier rubyid_create_and_add_features'><span class='object_link'><a href="#create_and_add_features-instance_method" title="#create_and_add_features (method)">create_and_add_features</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_bio_embl_as_biosequence'>bio_embl_as_biosequence</span><span class='comma'>,</span> <span class='id identifier rubyid_gene_id'>gene_id</span><span class='comma'>,</span> <span class='id identifier rubyid_all_positions_reverse'>all_positions_reverse</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>-</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
        <span class='comment'># adding the biosequence object to the new hash unless no new features were added
</span>        <span class='kw'>unless</span> <span class='id identifier rubyid_all_positions_forward'>all_positions_forward</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_all_positions_reverse'>all_positions_reverse</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> 
            <span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='lbracket'>[</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='id identifier rubyid_bio_embl_as_biosequence'>bio_embl_as_biosequence</span> <span class='kw'>unless</span> <span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='period'>.</span><span class='id identifier rubyid_keys'>keys</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rparen'>)</span>
        <span class='kw'>end</span>
    <span class='kw'>end</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span> <span class='comment'># return the new embl hash
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="get_embl-instance_method">
  
    #<strong>get_embl</strong>(gene_array)  &#x21d2; <tt>Hash&lt;Symbol, Bio::EMBL&gt;</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Given an array of gene ids, it gets the EMBL entry from ensemblgenomesgene,
creates a Bio::EMBL object and stores it into a hash</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>gene_array</span>
      
      
        <span class='type'>(<tt>Array&lt;Symbol&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The array of gene ids</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::EMBL&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The hash of Bio::EMBL objects for each EMBL entry</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 72</span>

<span class='kw'>def</span> <span class='id identifier rubyid_get_embl'>get_embl</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_array'>gene_array</span><span class='rparen'>)</span>
    <span class='kw'>unless</span> <span class='id identifier rubyid_gene_array'>gene_array</span> <span class='op'>&amp;&amp;</span> <span class='op'>!</span><span class='id identifier rubyid_gene_array'>gene_array</span><span class='period'>.</span><span class='id identifier rubyid_empty?'>empty?</span> <span class='comment'># if there is nothing in the array, warning and exit the function, returning nil
</span>        <span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>WARNING: no genes in the gene array / no gene array</span><span class='tstring_end'>&quot;</span></span>
        <span class='kw'>return</span> <span class='kw'>nil</span>
    <span class='kw'>end</span>
    <span class='comment'># else 
</span>    <span class='id identifier rubyid_embl_hash'>embl_hash</span> <span class='op'>=</span> <span class='lbrace'>{</span><span class='rbrace'>}</span> <span class='comment'># Hash to store {gene_id =&gt; Bio::EMBL object}
</span>    <span class='id identifier rubyid_gene_array'>gene_array</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_gene_array'>gene_array</span><span class='rbracket'>]</span> <span class='kw'>unless</span> <span class='id identifier rubyid_gene_array'>gene_array</span><span class='period'>.</span><span class='id identifier rubyid_is_a?'>is_a?</span><span class='lparen'>(</span><span class='const'>Array</span><span class='rparen'>)</span> <span class='comment'># converting it into an array if it isn&#39;t (if there&#39;s only one element)
</span>    <span class='id identifier rubyid_gene_array'>gene_array</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='op'>|</span> <span class='comment'># for each gene in the array
</span>        <span class='comment'># Getting an EMBL entry from dbfetch using the database ensemblgenomesgene (that accepts AGI locus codes)
</span>        <span class='id identifier rubyid_url'>url</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&amp;format=embl&amp;id=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='embexpr_end'>}</span><span class='tstring_content'>&amp;style=raw</span><span class='tstring_end'>&quot;</span></span>
        <span class='id identifier rubyid_response'>response</span> <span class='op'>=</span> <span class='id identifier rubyid_fetch'><span class='object_link'><a href="#fetch-instance_method" title="#fetch (method)">fetch</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_url'>url</span><span class='rparen'>)</span> <span class='comment'># accessing the url
</span>        <span class='kw'>if</span> <span class='id identifier rubyid_response'>response</span>
            <span class='id identifier rubyid_entry'>entry</span> <span class='op'>=</span> <span class='const'>Bio</span><span class='op'>::</span><span class='const'>EMBL</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_response'>response</span><span class='period'>.</span><span class='id identifier rubyid_body'>body</span><span class='rparen'>)</span> <span class='comment'># pass it the record as a string
</span>            <span class='kw'>unless</span> <span class='id identifier rubyid_embl_hash'>embl_hash</span><span class='period'>.</span><span class='id identifier rubyid_keys'>keys</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rparen'>)</span>
                <span class='id identifier rubyid_embl_hash'>embl_hash</span><span class='lbracket'>[</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rbracket'>]</span> <span class='op'>=</span> <span class='id identifier rubyid_entry'>entry</span> <span class='comment'># storing it in the hash
</span>            <span class='kw'>else</span>
                <span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>WARNING: </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='embexpr_end'>}</span><span class='tstring_content'> already contains an entry in the embl_hash</span><span class='tstring_end'>&quot;</span></span>
            <span class='kw'>end</span>
        <span class='kw'>end</span>
    <span class='kw'>end</span>
    <span class='kw'>return</span> <span class='id identifier rubyid_embl_hash'>embl_hash</span> <span class='comment'># returns the hash
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="read_gene_list-instance_method">
  
    #<strong>read_gene_list</strong>(path)  &#x21d2; <tt>Array&lt;Symbol&gt;</tt><sup>?</sup> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Reads a file containing one gene id (in AGI locus code format) in each
column and stores each id as a lowercase symbol on an array and returns it.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>path</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>Path of the txt file</p>
</div>
      
    </li>
  
</ul>

<p class="tag_title">Returns:</p>
<ul class="return">
  
    <li>
      
      
        <span class='type'>(<tt>Array&lt;Symbol&gt;</tt>, <tt>nil</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The array of ids, or nil if an error occured</p>
</div>
      
    </li>
  
</ul>

</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 45</span>

<span class='kw'>def</span> <span class='id identifier rubyid_read_gene_list'>read_gene_list</span><span class='lparen'>(</span><span class='id identifier rubyid_path'>path</span><span class='rparen'>)</span>
    <span class='comment'># for each gene in the file, stores the id as a lowercase symbol in gene_array (if it doesn&#39;t exist already)
</span>    <span class='id identifier rubyid_gene_array'>gene_array</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='rbracket'>]</span>
    <span class='kw'>begin</span>
        <span class='const'>IO</span><span class='period'>.</span><span class='id identifier rubyid_foreach'>foreach</span><span class='lparen'>(</span><span class='id identifier rubyid_path'>path</span><span class='rparen'>)</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene'>gene</span><span class='op'>|</span>
            <span class='id identifier rubyid_gene'>gene</span><span class='period'>.</span><span class='id identifier rubyid_strip!'>strip!</span><span class='period'>.</span><span class='id identifier rubyid_downcase!'>downcase!</span> <span class='comment'># removing whitespace (if any) and converting the letters to lowercase
</span>            <span class='id identifier rubyid_gene_array'>gene_array</span> <span class='op'>|=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_gene'>gene</span><span class='period'>.</span><span class='id identifier rubyid_to_sym'>to_sym</span><span class='rbracket'>]</span> <span class='kw'>if</span> <span class='id identifier rubyid_gene'>gene</span> <span class='op'>=~</span> <span class='tstring'><span class='regexp_beg'>/</span><span class='tstring_content'>at\wg\d\d\d\d\d</span><span class='regexp_end'>/</span></span> <span class='comment'># storing the gene id as a symbol if it matches the regexp
</span>        <span class='kw'>end</span>
        <span class='kw'>return</span> <span class='id identifier rubyid_gene_array'>gene_array</span> <span class='comment'># returns the gene_array
</span>    <span class='kw'>rescue</span> <span class='const'>Errno</span><span class='op'>::</span><span class='const'>ENOENT</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span> <span class='comment'># handling the missing file exception
</span>        <span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ERROR: Can&#39;t find the file</span><span class='tstring_end'>&quot;</span></span>
        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='kw'>return</span> <span class='kw'>nil</span>
    <span class='kw'>rescue</span> <span class='const'>Exception</span> <span class='op'>=&gt;</span> <span class='id identifier rubyid_e'>e</span> <span class='comment'># other possible exceptions
</span>        <span class='gvar'>$stderr</span><span class='period'>.</span><span class='id identifier rubyid_puts'>puts</span> <span class='id identifier rubyid_e'>e</span><span class='period'>.</span><span class='id identifier rubyid_inspect'>inspect</span>
        <span class='kw'>return</span> <span class='kw'>nil</span>
    <span class='kw'>end</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_gff3-instance_method">
  
    #<strong>write_gff3</strong>(new_embl_hash, mode, filename)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Writes a GFF3 file with all of the &#39;CTTCTT_direct_repeat&#39; features.
It has two modes, “local” uses the start of the gene as a coordinate
origin,  and “global” uses the start of the chromosome.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>new_embl_hash</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The hash of Bio::Sequence objects, containing the
&#39;CTTCTT_direct_repeat&#39; features</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>mode</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The mode for the coordinate system, “local” or “global”. If none of those,
the default is “local”</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>filename</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The name of the output file</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277
278
279
280
281
282
283
284
285
286
287
288
289
290
291
292
293
294
295
296
297
298
299
300
301
302
303
304
305
306
307
308
309</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 257</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_gff3'>write_gff3</span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='comma'>,</span> <span class='id identifier rubyid_mode'>mode</span><span class='comma'>,</span> <span class='id identifier rubyid_filename'>filename</span><span class='rparen'>)</span>

    <span class='comment'>#seqid # id of the landmark: gene id in local, chr number in global
</span>    <span class='comment'>#source = &quot;BioRuby&quot;
</span>    <span class='comment'>#type = &quot;direct_repeat&quot; # SO:0000314, direct repeat because in CTTCTT, CTT is repeated in the same orientation
</span>    <span class='comment'>#start_pos # relative to the gene id in local, relative to the chr in global
</span>    <span class='comment'>#end_pos # same as above
</span>    <span class='comment'>#score = &quot;.&quot;
</span>    <span class='comment'>#strand # strand relative to the landmark
</span>    <span class='comment'>#phase = &quot;.&quot; # not a CDS
</span>    <span class='comment'>#attributes # ID=unique_id;Name=CTTCTT_direct_repeat;
</span>
    <span class='id identifier rubyid_puts'>puts</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>WARNING: mode isn&#39;t &#39;local&#39; or &#39;global&#39;, the default &#39;local&#39; will be used</span><span class='tstring_end'>&quot;</span></span> <span class='kw'>if</span> <span class='id identifier rubyid_mode'>mode</span> <span class='op'>!=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>local</span><span class='tstring_end'>&quot;</span></span> <span class='op'>&amp;&amp;</span> <span class='id identifier rubyid_mode'>mode</span> <span class='op'>!=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>global</span><span class='tstring_end'>&quot;</span></span>

    <span class='id identifier rubyid_source'>source</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>BioRuby</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_type'>type</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>direct_repeat</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_score'>score</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>
    <span class='id identifier rubyid_phase'>phase</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>.</span><span class='tstring_end'>&quot;</span></span>

    <span class='id identifier rubyid_f'>f</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_filename'>filename</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>##gff-version 3\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='comment'># writing the first line
</span>    <span class='comment'># iterating through every biosequence object
</span>    <span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='comma'>,</span> <span class='id identifier rubyid_biosequence'>biosequence</span><span class='op'>|</span>
        <span class='id identifier rubyid_biosequence'>biosequence</span><span class='period'>.</span><span class='id identifier rubyid_features'>features</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_feature'>feature</span><span class='op'>|</span>
            <span class='kw'>next</span> <span class='kw'>unless</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_direct_repeat</span><span class='tstring_end'>&quot;</span></span> <span class='comment'># skip if it is not this feature
</span>
            <span class='comment'># getting the rest of the elements of the gff3
</span>            <span class='id identifier rubyid_bio_location'>bio_location</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_locations'>locations</span><span class='lbracket'>[</span><span class='int'>0</span><span class='rbracket'>]</span>
            <span class='comment'># unique elements to each gff3 file
</span>            <span class='kw'>if</span> <span class='id identifier rubyid_mode'>mode</span> <span class='op'>==</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>global</span><span class='tstring_end'>&quot;</span></span>
                <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Chr</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='lbracket'>[</span><span class='int'>2</span><span class='rbracket'>]</span><span class='embexpr_end'>}</span><span class='tstring_end'>&quot;</span></span>
                <span class='id identifier rubyid_chr_start_pos'>chr_start_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_biosequence'>biosequence</span><span class='period'>.</span><span class='id identifier rubyid_primary_accession'>primary_accession</span><span class='period'>.</span><span class='id identifier rubyid_split'>split</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>:</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span><span class='lbracket'>[</span><span class='int'>3</span><span class='rbracket'>]</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='comment'># chromosome:TAIR10:3:20119140:20121314:1 -&gt; [3] is 20119140
</span>                <span class='comment'># for the positions, both the gene position in the chromosome and the feature position in the gene are 1-indexed
</span>                <span class='comment'># so we have to substract 1 to the sum
</span>                <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_chr_start_pos'>chr_start_pos</span> <span class='op'>+</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span> 
                <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='lparen'>(</span><span class='id identifier rubyid_chr_start_pos'>chr_start_pos</span> <span class='op'>+</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span><span class='period'>.</span><span class='id identifier rubyid_to_i'>to_i</span> <span class='op'>-</span> <span class='int'>1</span><span class='rparen'>)</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span> 
            <span class='kw'>else</span> <span class='comment'># default and if mode == &quot;local&quot;
</span>                <span class='id identifier rubyid_seqid'>seqid</span> <span class='op'>=</span> <span class='id identifier rubyid_gene_id'>gene_id</span><span class='period'>.</span><span class='id identifier rubyid_to_s'>to_s</span><span class='period'>.</span><span class='id identifier rubyid_upcase'>upcase</span>
                <span class='id identifier rubyid_start_pos'>start_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_from'>from</span>
                <span class='id identifier rubyid_end_pos'>end_pos</span> <span class='op'>=</span> <span class='id identifier rubyid_bio_location'>bio_location</span><span class='period'>.</span><span class='id identifier rubyid_to'>to</span>
            <span class='kw'>end</span>
            <span class='comment'># common elements in both gff3 files
</span>            <span class='id identifier rubyid_strand'>strand</span> <span class='op'>=</span> <span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>strand</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span>
            <span class='id identifier rubyid_attributes'>attributes</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_assoc'>assoc</span><span class='lbracket'>[</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>ID</span><span class='tstring_end'>&quot;</span></span><span class='rbracket'>]</span><span class='embexpr_end'>}</span><span class='tstring_content'>;Name=</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_feature'>feature</span><span class='period'>.</span><span class='id identifier rubyid_feature'>feature</span><span class='embexpr_end'>}</span><span class='tstring_content'>;</span><span class='tstring_end'>&quot;</span></span>

            <span class='comment'># joining the elements and writing them to the file
</span>            <span class='id identifier rubyid_elements'>elements</span> <span class='op'>=</span> <span class='lbracket'>[</span><span class='id identifier rubyid_seqid'>seqid</span><span class='comma'>,</span> <span class='id identifier rubyid_source'>source</span><span class='comma'>,</span> <span class='id identifier rubyid_type'>type</span><span class='comma'>,</span> <span class='id identifier rubyid_start_pos'>start_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_end_pos'>end_pos</span><span class='comma'>,</span> <span class='id identifier rubyid_score'>score</span><span class='comma'>,</span> <span class='id identifier rubyid_strand'>strand</span><span class='comma'>,</span> <span class='id identifier rubyid_phase'>phase</span><span class='comma'>,</span> <span class='id identifier rubyid_attributes'>attributes</span><span class='rbracket'>]</span>
            <span class='id identifier rubyid_entry'>entry</span> <span class='op'>=</span> <span class='id identifier rubyid_elements'>elements</span><span class='period'>.</span><span class='id identifier rubyid_join'>join</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\t</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='op'>+</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span>
            <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='id identifier rubyid_entry'>entry</span><span class='rparen'>)</span>
        <span class='kw'>end</span>
    <span class='kw'>end</span>
    <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_close'>close</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_gff3_global-instance_method">
  
    #<strong>write_gff3_global</strong>(new_embl_hash, filename = &quot;CTTCTT_GFF3_chromosome.gff&quot;)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Writes a GFF3 file with all of the &#39;CTTCTT_direct_repeat&#39; features,
using the start of the chromosome as the coordinate origin.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>new_embl_hash</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The hash of Bio::Sequence objects, containing the
&#39;CTTCTT_direct_repeat&#39; features</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>filename</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;CTTCTT_GFF3_chromosome.gff&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>The name of the output file</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


243
244
245</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 243</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_gff3_global'>write_gff3_global</span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='comma'>,</span> <span class='id identifier rubyid_filename'>filename</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_GFF3_chromosome.gff</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_write_gff3'><span class='object_link'><a href="#write_gff3-instance_method" title="#write_gff3 (method)">write_gff3</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='comma'>,</span> <span class='id identifier rubyid_mode'>mode</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>global</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_filename'>filename</span> <span class='op'>=</span> <span class='id identifier rubyid_filename'>filename</span><span class='rparen'>)</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_gff3_local-instance_method">
  
    #<strong>write_gff3_local</strong>(new_embl_hash, filename = &quot;CTTCTT_GFF3_gene.gff&quot;)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Writes a GFF3 file with all of the &#39;CTTCTT_direct_repeat&#39; features,
using the start of the gene as the coordinate origin.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>new_embl_hash</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The hash of Bio::Sequence objects, containing the
&#39;CTTCTT_direct_repeat&#39; features</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>filename</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;CTTCTT_GFF3_gene.gff&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>The name of the output file</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


233
234
235</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 233</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_gff3_local'>write_gff3_local</span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='comma'>,</span> <span class='id identifier rubyid_filename'>filename</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_GFF3_gene.gff</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_write_gff3'><span class='object_link'><a href="#write_gff3-instance_method" title="#write_gff3 (method)">write_gff3</a></span></span><span class='lparen'>(</span><span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='comma'>,</span> <span class='id identifier rubyid_mode'>mode</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>local</span><span class='tstring_end'>&quot;</span></span><span class='comma'>,</span> <span class='id identifier rubyid_filename'>filename</span> <span class='op'>=</span> <span class='id identifier rubyid_filename'>filename</span><span class='rparen'>)</span>
<span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
      <div class="method_details ">
  <h3 class="signature " id="write_report-instance_method">
  
    #<strong>write_report</strong>(gene_array, new_embl_hash, filename = &quot;CTTCTT_report.txt&quot;)  &#x21d2; <tt>Object</tt> 
  

  

  
</h3><div class="docstring">
  <div class="discussion">
    
<p>Writes a report with the IDs of all the genes in the initial gene list that
don&#39;t contain &#39;CTTCTT&#39; in any of their exons.</p>


  </div>
</div>
<div class="tags">
  <p class="tag_title">Parameters:</p>
<ul class="param">
  
    <li>
      
        <span class='name'>gene_array</span>
      
      
        <span class='type'>(<tt>Array&lt;Symbol&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The array of gene ids</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>new_embl_hash</span>
      
      
        <span class='type'>(<tt>Hash&lt;Symbol, Bio::Sequence&gt;</tt>)</span>
      
      
      
        &mdash;
        <div class='inline'>
<p>The hash of Bio::Sequence objects, containing the
&#39;CTTCTT_direct_repeat&#39; features</p>
</div>
      
    </li>
  
    <li>
      
        <span class='name'>filename</span>
      
      
        <span class='type'>(<tt>String</tt>)</span>
      
      
        <em class="default">(defaults to: <tt>&quot;CTTCTT_report.txt&quot;</tt>)</em>
      
      
        &mdash;
        <div class='inline'>
<p>The name of the output file</p>
</div>
      
    </li>
  
</ul>


</div><table class="source_code">
  <tr>
    <td>
      <pre class="lines">


319
320
321
322
323
324
325
326
327
328
329
330</pre>
    </td>
    <td>
      <pre class="code"><span class="info file"># File 'assignment3_functions.rb', line 319</span>

<span class='kw'>def</span> <span class='id identifier rubyid_write_report'>write_report</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_array'>gene_array</span><span class='comma'>,</span> <span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='comma'>,</span> <span class='id identifier rubyid_filename'>filename</span> <span class='op'>=</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>CTTCTT_report.txt</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='comment'># getting the genes from gene_array that aren&#39;t in new_embl_hash
</span>    <span class='id identifier rubyid_not_feature_genes'>not_feature_genes</span> <span class='op'>=</span> <span class='id identifier rubyid_gene_array'>gene_array</span><span class='period'>.</span><span class='id identifier rubyid_select'>select</span> <span class='lbrace'>{</span><span class='op'>|</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='op'>|</span> <span class='op'>!</span><span class='id identifier rubyid_new_embl_hash'>new_embl_hash</span><span class='period'>.</span><span class='id identifier rubyid_keys'>keys</span><span class='period'>.</span><span class='id identifier rubyid_include?'>include?</span><span class='lparen'>(</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='rparen'>)</span><span class='rbrace'>}</span> 
    <span class='id identifier rubyid_f'>f</span> <span class='op'>=</span> <span class='const'>File</span><span class='period'>.</span><span class='id identifier rubyid_new'>new</span><span class='lparen'>(</span><span class='id identifier rubyid_filename'>filename</span><span class='comma'>,</span> <span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>w</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span> <span class='comment'># opening the file
</span>    <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Number of genes present in the initial txt: </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_array'>gene_array</span><span class='period'>.</span><span class='id identifier rubyid_length'>length</span><span class='embexpr_end'>}</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>Number of genes without CTTCTT in any of their exons: </span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_not_feature_genes'>not_feature_genes</span><span class='period'>.</span><span class='id identifier rubyid_length'>length</span><span class='embexpr_end'>}</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='tstring_content'>List of genes: \n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='id identifier rubyid_not_feature_genes'>not_feature_genes</span><span class='period'>.</span><span class='id identifier rubyid_each'>each</span> <span class='kw'>do</span> <span class='op'>|</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='op'>|</span>
        <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_write'>write</span><span class='lparen'>(</span><span class='tstring'><span class='tstring_beg'>&quot;</span><span class='embexpr_beg'>#{</span><span class='id identifier rubyid_gene_id'>gene_id</span><span class='embexpr_end'>}</span><span class='tstring_content'>\n</span><span class='tstring_end'>&quot;</span></span><span class='rparen'>)</span>
    <span class='kw'>end</span>
    <span class='id identifier rubyid_f'>f</span><span class='period'>.</span><span class='id identifier rubyid_close'>close</span> <span class='comment'># closing the file
</span><span class='kw'>end</span></pre>
    </td>
  </tr>
</table>
</div>
    
  </div>

</div>

      <div id="footer">
  Generated on Sun Nov  7 15:44:28 2021 by
  <a href="http://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.25 (ruby-2.5.1).
</div>

    </div>
  </body>
</html>