<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>
  File: README
  
    &mdash; Documentation by YARD 0.9.25
  
</title>

  <link rel="stylesheet" href="css/style.css" type="text/css" />

  <link rel="stylesheet" href="css/common.css" type="text/css" />

<script type="text/javascript">
  pathId = "README";
  relpath = '';
</script>


  <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

  <script type="text/javascript" charset="utf-8" src="js/app.js"></script>


  </head>
  <body>
    <div class="nav_wrap">
      <iframe id="nav" src="class_list.html?1"></iframe>
      <div id="resizer"></div>
    </div>

    <div id="main" tabindex="-1">
      <div id="header">
        <div id="menu">
  
    <a href="_index.html">Index</a> &raquo; 
    <span class="title">File: README</span>
  
</div>

        <div id="search">
  
    <a class="full_list_link" id="class_list_link"
        href="class_list.html">

        <svg width="24" height="24">
          <rect x="0" y="4" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="12" width="24" height="4" rx="1" ry="1"></rect>
          <rect x="0" y="20" width="24" height="4" rx="1" ry="1"></rect>
        </svg>
    </a>
  
</div>
        <div class="clear"></div>
      </div>

      <div id="content"><div id='filecontents'>
<p>Assignment 4. Command to run:</p>

<p>ruby main.rb ./blast_databases/schizosaccharomyces_pombe.fa
./blast_databases/arabidopsis_thaliana.fa</p>

<p>The results of the analysis are in BRH_report.tsv</p>
<hr>

<p>About the fasta files:</p>

<p>In this case, the fasta files provided for Arabidopsis and S. pombe
don&#39;t have the problem we mentioned in class about the non-unique
identifiers.</p>
<hr>

<p>About the parameters:</p>

<p>From <a
href="https://doi.org/10.1093/bioinformatics/btm585">doi.org/10.1093/bioinformatics/btm585</a></p>

<p>They recommend an E-value threshold of 1*10^-6 and that there is a query
coverage of at least 50%.</p>

<p>Also, the best detection of orthologs as best reciprocal hits was obtained
with soft filtering and a Smith-Waterman final alignment (-F “m S” -s T),
giving both the highest number of orthologs and the minimal error rates.
However, using a Smith-Waterman final alignment is computationally
expensive, and almost the same results were achieved just by using soft
filtering (-F “m S”), which is what I&#39;ll use.</p>
<hr>

<p>Next steps in the search of orthologs:</p>

<p>From <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3193375">www.ncbi.nlm.nih.gov/pmc/articles/PMC3193375</a>/</p>

<p>A reasonable next step would be to include a third species, and perform a
COG search (cluster of orthologous genes), a clustering method based on
best reciprocal hit. We could identify the best reciprocal hits between the
three species, and find the clusters that contain a best reciprocal hit
from the three species. Because we have included one additional species, it
is more probable that the orthologous genes we find here are really
orthologous.</p>

<p>Furthermore, we could use tree-based approaches, building phylogenetic gene
trees between three or more species, and looking for orthologous genes
between our species of interest.</p>
</div></div>

      <div id="footer">
  Generated on Wed Dec  1 11:00:29 2021 by
  <a href="http://yardoc.org" title="Yay! A Ruby Documentation Tool" target="_parent">yard</a>
  0.9.25 (ruby-2.5.1).
</div>

    </div>
  </body>
</html>